src/mscorlib/src/System/Reflection/CustomAttribute.cs

```diff
@@ -1982,9 +1979,9 @@ internal static Attribute[] GetCustomAttributes(RuntimeMethodInfo method, Runtim
 
             bool all = caType == (RuntimeType)typeof(object) || caType == (RuntimeType)typeof(Attribute);
             if (!all && !s_pca.ContainsKey(caType))
-                return Array.Empty<Attribute>();
+                return null;
 
-            List<Attribute> pcas = new List<Attribute>();
+            Attribute[] pcas = new Attribute[all ? 2 : 1];
             Attribute pca;
 
             if (all || caType == (RuntimeType)typeof(DllImportAttribute))
```

Comment:
The actual bug was on `if (pca != null) pcas[count++] = pca` line. We were indexing beyond end of the list.

I have fixed this by consistently using preallocated array everywhere. The reason why some methods used preallocated array and other List does not hold anymore.